File: packages/sdk-go/cache_test.go
package rollgate

import (
	"testing"
	"time"
)

func TestFlagCache_GetSet(t *testing.T) {
	t.Run("should return empty result for empty cache", func(t *testing.T) {
		cache := NewFlagCache(DefaultCacheConfig())
		result := cache.Get()

		if result.Found {
			t.Error("expected Found to be false")
		}
	})

	t.Run("should store and retrieve flags", func(t *testing.T) {
		cache := NewFlagCache(DefaultCacheConfig())
		flags := map[string]bool{"test-flag": true, "another-flag": false}

		cache.Set(flags)
		result := cache.Get()

		if !result.Found {
			t.Error("expected Found to be true")
		}
		if result.Stale {
			t.Error("expected Stale to be false")
		}
		if result.Flags["test-flag"] != true {
			t.Error("expected test-flag to be true")
		}
		if result.Flags["another-flag"] != false {
			t.Error("expected another-flag to be false")
		}
	})
}

func TestFlagCache_TTL(t *testing.T) {
	t.Run("should return stale data after TTL", func(t *testing.T) {
		config := CacheConfig{
			TTL:      10 * time.Millisecond,
			StaleTTL: 1 * time.Hour,
			Enabled:  true,
		}
		cache := NewFlagCache(config)
		cache.Set(map[string]bool{"test": true})

		// Wait for TTL to expire
		time.Sleep(15 * time.Millisecond)

		result := cache.Get()
		if !result.Found {
			t.Error("expected Found to be true")
		}
		if !result.Stale {
			t.Error("expected Stale to be true")
		}
	})

	t.Run("should return empty after staleTTL", func(t *testing.T) {
		config := CacheConfig{
			TTL:      5 * time.Millisecond,
			StaleTTL: 10 * time.Millisecond,
			Enabled:  true,
		}
		cache := NewFlagCache(config)
		cache.Set(map[string]bool{"test": true})

		// Wait for staleTTL to expire
		time.Sleep(15 * time.Millisecond)

		result := cache.Get()
		if result.Found {
			t.Error("expected Found to be false")
		}
	})
}

func TestFlagCache_HasFreshHasAny(t *testing.T) {
	t.Run("HasFresh should return true for fresh data", func(t *testing.T) {
		cache := NewFlagCache(DefaultCacheConfig())
		cache.Set(map[string]bool{"test": true})

		if !cache.HasFresh() {
			t.Error("expected HasFresh to be true")
		}
	})

	t.Run("HasAny should return true for stale data", func(t *testing.T) {
		config := CacheConfig{
			TTL:      5 * time.Millisecond,
			StaleTTL: 1 * time.Hour,
			Enabled:  true,
		}
		cache := NewFlagCache(config)
		cache.Set(map[string]bool{"test": true})

		time.Sleep(10 * time.Millisecond)

		if cache.HasFresh() {
			t.Error("expected HasFresh to be false")
		}
		if !cache.HasAny() {
			t.Error("expected HasAny to be true")
		}
	})

	t.Run("both should return false for empty cache", func(t *testing.T) {
		cache := NewFlagCache(DefaultCacheConfig())

		if cache.HasFresh() {
			t.Error("expected HasFresh to be false")
		}
		if cache.HasAny() {
			t.Error("expected HasAny to be false")
		}
	})
}

func TestFlagCache_Clear(t *testing.T) {
	cache := NewFlagCache(DefaultCacheConfig())
	cache.Set(map[string]bool{"test": true})
	cache.Clear()

	result := cache.Get()
	if result.Found {
		t.Error("expected Found to be false after Clear")
	}
}

func TestFlagCache_Stats(t *testing.T) {
	t.Run("should track hits", func(t *testing.T) {
		cache := NewFlagCache(DefaultCacheConfig())
		cache.Set(map[string]bool{"test": true})

		cache.Get()
		cache.Get()

		stats := cache.GetStats()
		if stats.Hits != 2 {
			t.Errorf("expected 2 hits, got %d", stats.Hits)
		}
	})

	t.Run("should track misses", func(t *testing.T) {
		cache := NewFlagCache(DefaultCacheConfig())

		cache.Get()
		cache.Get()

		stats := cache.GetStats()
		if stats.Misses != 2 {
			t.Errorf("expected 2 misses, got %d", stats.Misses)
		}
	})

	t.Run("should track stale hits", func(t *testing.T) {
		config := CacheConfig{
			TTL:      5 * time.Millisecond,
			StaleTTL: 1 * time.Hour,
			Enabled:  true,
		}
		cache := NewFlagCache(config)
		cache.Set(map[string]bool{"test": true})

		time.Sleep(10 * time.Millisecond)

		cache.Get()
		cache.Get()

		stats := cache.GetStats()
		if stats.StaleHits != 2 {
			t.Errorf("expected 2 stale hits, got %d", stats.StaleHits)
		}
	})
}

File: packages/sdk-core/src/cache.test.ts
import { describe, it, expect, beforeEach, afterEach, vi } from "vitest";
import { FlagCache, DEFAULT_CACHE_CONFIG } from "./cache";

describe("FlagCache", () => {
  beforeEach(() => {
    vi.useFakeTimers();
    if (typeof localStorage !== "undefined") {
      localStorage.clear();
    }
  });

  afterEach(() => {
    vi.useRealTimers();
  });

  describe("DEFAULT_CACHE_CONFIG", () => {
    it("should have sensible defaults", () => {
      expect(DEFAULT_CACHE_CONFIG.ttl).toBe(300000);
      expect(DEFAULT_CACHE_CONFIG.staleTtl).toBe(3600000);
    });
  });

  describe("get/set", () => {
    it("should return undefined for empty cache", () => {
      const cache = new FlagCache();
      expect(cache.get()).toBeUndefined();
    });

    it("should store and retrieve flags", () => {
      const cache = new FlagCache();
      const flags = { "test-flag": true, "another-flag": false };

      cache.set(flags);
      const result = cache.get();

      expect(result).toBeDefined();
      expect(result?.flags).toEqual(flags);
      expect(result?.stale).toBe(false);
    });

    it("should return fresh data within TTL", () => {
      const cache = new FlagCache({ ttl: 5000 });
      const flags = { test: true };

      cache.set(flags);
      vi.advanceTimersByTime(3000);

      const result = cache.get();
      expect(result?.stale).toBe(false);
    });

    it("should return stale data after TTL but within staleTtl", () => {
      const cache = new FlagCache({ ttl: 5000, staleTtl: 60000 });
      const flags = { test: true };

      cache.set(flags);
      vi.advanceTimersByTime(10000);

      const result = cache.get();
      expect(result).toBeDefined();
      expect(result?.stale).toBe(true);
    });

    it("should return undefined after staleTtl", () => {
      const cache = new FlagCache({ ttl: 5000, staleTtl: 60000 });
      const flags = { test: true };

      cache.set(flags);
      vi.advanceTimersByTime(70000);

      const result = cache.get();
      expect(result).toBeUndefined();
    });
  });

  describe("hasFresh/hasAny", () => {
    it("hasFresh should return true only for fresh data", () => {
      const cache = new FlagCache({ ttl: 5000 });
      cache.set({ test: true });

      expect(cache.hasFresh()).toBe(true);

      vi.advanceTimersByTime(6000);
      expect(cache.hasFresh()).toBe(false);
    });

    it("hasAny should return true for stale data", () => {
      const cache = new FlagCache({ ttl: 5000, staleTtl: 60000 });
      cache.set({ test: true });

      vi.advanceTimersByTime(10000);

      expect(cache.hasFresh()).toBe(false);
      expect(cache.hasAny()).toBe(true);
    });

    it("hasAny should return false after staleTtl", () => {
      const cache = new FlagCache({ ttl: 5000, staleTtl: 60000 });
      cache.set({ test: true });

      vi.advanceTimersByTime(70000);

      expect(cache.hasAny()).toBe(false);
    });
  });

  describe("clear", () => {
    it("should remove all cached data", () => {
      const cache = new FlagCache();
      cache.set({ test: true });

      cache.clear();

      expect(cache.get()).toBeUndefined();
    });
  });

  describe("stats", () => {
    it("should track cache hits", () => {
      const cache = new FlagCache();
      cache.set({ test: true });

      cache.get();
      cache.get();

      const stats = cache.getStats();
      expect(stats.hits).toBe(2);
    });

    it("should track cache misses", () => {
      const cache = new FlagCache();

      cache.get();
      cache.get();

      const stats = cache.getStats();
      expect(stats.misses).toBe(2);
    });

    it("should track stale hits", () => {
      const cache = new FlagCache({ ttl: 1000, staleTtl: 60000 });
      cache.set({ test: true });

      vi.advanceTimersByTime(2000);

      cache.get();
      cache.get();

      const stats = cache.getStats();
      expect(stats.staleHits).toBe(2);
    });
  });
});
